core/store/src/db.rs

```diff
@@ -79,6 +80,119 @@ impl DBTransaction {
 
 pub type DBIterator<'a> = Box<dyn Iterator<Item = io::Result<(Box<[u8]>, Box<[u8]>)>> + 'a>;
 
+/// Data returned from the database.
+///
+/// Abstraction layer for data returned by [`Database::get_raw_bytes`] and
+/// [`Database::get_with_rc_stripped`] methods.  Operating on the value as
+/// a slice is free while converting it to a vector on an arc may requires an
+/// allocation and memory copy.
+pub struct DBBytes<'a>(DBBytesInner<'a>);
+
+enum DBBytesInner<'a> {
+    /// Data held as a vector.
+    Vec(Vec<u8>),
+
+    /// Data held as RocksDB-specific pinnable slice.
+    Rocks {
+        /// Slice view of the data.
+        ///
+        /// Having this allows us to create views at only parts of the pinnable
+        /// slice without having to modify it.  We need it because remove_suffix
+        /// method of the C++ class is not exposed.
+        slice: &'a [u8],
+
+        /// This is what owns the data.
+        ///
+        /// This is never modified which is why we can have `slice` field
+        /// pointing at the data this pinnable slice holds.
+        db_slice: ::rocksdb::DBPinnableSlice<'a>,
+    },
+}
+
+impl<'a> DBBytes<'a> {
+    pub fn as_slice(&self) -> &[u8] {
+        match self.0 {
+            DBBytesInner::Vec(ref bytes) => bytes.as_slice(),
+            DBBytesInner::Rocks { slice, .. } => slice,
+        }
+    }
+
+    fn from_rocksdb_slice(db_slice: ::rocksdb::DBPinnableSlice<'a>) -> Self {
+        let slice = &*db_slice;
+        let (ptr, len) = (slice.as_ptr(), slice.len());
+        // SAFETY: We got the raw parts from a slice so they are obviously
+        // correct.  By using 'a lifetime we guarantee that this slice won’t
+        // outlive `db_slice` field.  And lastly, since `db_slice` is never
```

Comment:
I'd suggest the following ways to make this better: 

* if we keep slice, let's move `DBBytes` to a separate module, so that the wrong lifetime is properly private (I'd also  move to a separate file anyway -- there's a bunch of boring impls here, no need to add noise to lib.rs)
* but I would say that using `*const [u8]` rather than a slice with a lifetime would be a less lying-to-the-compiler way to go here. That way, the `unsafe` would live in the `as_slice` method, where it belongs.
* alternatively, we can store just the offset, rather than the whole slice (which would also give us smaller size_of), and do slicing on access.

Reply:
Holding the whole slice rather than just length is actually a µoptimissation to avoid FFI calls each time we need to get the pointer out.   Since our other variant is Vec<u8> holding just the offset won’t actually reduce the size of the enum I don’t think.